File: agent/packages/schema/src/forge-bridge.ts
import { Type, type Static } from "@sinclair/typebox";

import { ColorKind } from "./cultivation.js";
import { ForgeOutcomeBucket } from "./forge.js";

const JS_SAFE_INTEGER_MAX = Number.MAX_SAFE_INTEGER;

export const ForgeMaterialStackV1 = Type.Object(
  {
    material: Type.String({ minLength: 1, maxLength: 128 }),
    count: Type.Integer({ minimum: 1, maximum: JS_SAFE_INTEGER_MAX }),
  },
  { additionalProperties: false },
);
export type ForgeMaterialStackV1 = Static<typeof ForgeMaterialStackV1>;

export const ForgeStartPayloadV1 = Type.Object(
  {
    v: Type.Literal(1),
    session_id: Type.Integer({ minimum: 0, maximum: JS_SAFE_INTEGER_MAX }),
    blueprint_id: Type.String({ minLength: 1, maxLength: 128 }),
    station_id: Type.String({ minLength: 1, maxLength: 128 }),
    caster_id: Type.String({ minLength: 1, maxLength: 128 }),
    materials: Type.Array(ForgeMaterialStackV1),
    ts: Type.Integer({ minimum: 0, maximum: JS_SAFE_INTEGER_MAX }),
  },
  { additionalProperties: false },
);
export type ForgeStartPayloadV1 = Static<typeof ForgeStartPayloadV1>;

export const ForgeOutcomePayloadV1 = Type.Object(
  {
    v: Type.Literal(1),
    session_id: Type.Integer({ minimum: 0, maximum: JS_SAFE_INTEGER_MAX }),
    blueprint_id: Type.String({ minLength: 1, maxLength: 128 }),
    bucket: ForgeOutcomeBucket,
    weapon_item: Type.Optional(Type.String({ minLength: 1, maxLength: 128 })),
    quality: Type.Number({ minimum: 0, maximum: 1 }),
    color: Type.Optional(ColorKind),
    side_effects: Type.Array(Type.String({ minLength: 1, maxLength: 128 })),
    achieved_tier: Type.Integer({ minimum: 0, maximum: 4 }),
    caster_id: Type.String({ minLength: 1, maxLength: 128 }),
    ts: Type.Integer({ minimum: 0, maximum: JS_SAFE_INTEGER_MAX }),
  },
  { additionalProperties: false },
);
export type ForgeOutcomePayloadV1 = Static<typeof ForgeOutcomePayloadV1>;

File: server/src/schema/forge_bridge.rs
//! 炼器（武器）server→agent Redis bridge payload。

use serde::{Deserialize, Serialize};

use super::forge::ForgeOutcomeBucketV1;
use crate::cultivation::components::ColorKind;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForgeMaterialStackV1 {
    pub material: String,
    pub count: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForgeStartPayloadV1 {
    pub v: u8,
    pub session_id: u64,
    pub blueprint_id: String,
    pub station_id: String,
    pub caster_id: String,
    pub materials: Vec<ForgeMaterialStackV1>,
    pub ts: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ForgeOutcomePayloadV1 {
    pub v: u8,
    pub session_id: u64,
    pub blueprint_id: String,
    pub bucket: ForgeOutcomeBucketV1,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub weapon_item: Option<String>,
    pub quality: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<ColorKind>,
    pub side_effects: Vec<String>,
    pub achieved_tier: u8,
    pub caster_id: String,
    pub ts: u64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn forge_start_payload_sample_roundtrip() {
        let payload: ForgeStartPayloadV1 = serde_json::from_str(include_str!(
            "../../../agent/packages/schema/samples/forge-start-payload.sample.json"
        ))
        .expect("forge start sample should deserialize");

        assert_eq!(payload.v, 1);
        assert_eq!(payload.blueprint_id, "qing_feng_v0");
        assert_eq!(payload.materials[0].material, "fan_tie");
    }

    #[test]
    fn forge_outcome_payload_samples_roundtrip() {
        for sample in [
            include_str!(
                "../../../agent/packages/schema/samples/forge-outcome-payload-perfect.sample.json"
            ),
            include_str!(
                "../../../agent/packages/schema/samples/forge-outcome-payload-flawed.sample.json"
            ),
        ] {
            let payload: ForgeOutcomePayloadV1 =
                serde_json::from_str(sample).expect("forge outcome sample should deserialize");
            assert_eq!(payload.v, 1);
            assert!((0.0..=1.0).contains(&payload.quality));
        }
    }
}
